Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/TapToPlace/TapToPlace.cs

```diff
@@ -0,0 +1,467 @@
+        /// </summary>
+        public bool IsBeingPlaced { get; protected set; }
+
+        [SerializeField]
+        [Tooltip("If true, the gameObjectToPlace will start in the placing state.  Meaning that the object will immediately start" +
+        "by following the TrackedTargetType (Head or Controller Ray) and then a tap is required to place the object.  AutoStart" +
+         "is only called once because it is in Start();")]
+        private bool autoStart = false;
+
+        /// <summary>
+        /// If true, the gameObjectToPlace will start in the placing state.  Meaning that the object will immediately start
+        /// by following the TrackedTargetType (Head or Controller Ray) and then a tap is required to place the object.  AutoStart
+        /// is only called once because it is in Start();
+        /// </summary>
+        public bool AutoStart
+        {
+            get => autoStart;
+            set
+            {
+                if (autoStart != value)
+                {
+                    autoStart = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// If true, the raycast did hit a surface
+        /// </summary>
+        public bool DidHit { get; protected set; }
+
+        [SerializeField]
+        [Tooltip("The distance between the center of the gameobject to place and a surface along the surface normal, if the raycast hits a surface")]
+        private float surfaceNormalOffset = 0.0f;
+
+        /// <summary>
+        /// SurfaceNormalOffset is the offset between the game object's center and The default value for SurfaceNormalOffset is half the depth of a game object. 
+        /// </summary>
+        public float SurfaceNormalOffset
+        {
+            get => surfaceNormalOffset;
+            set
+            {
+                if (surfaceNormalOffset != value)
+                {
+                    surfaceNormalOffset = value;
+                }
+            }
+        }
+
+        [SerializeField]
+        [Tooltip("If true, the GameObjectToPlace will remain upright and in line with Vector3.up")]
+        private bool keepOrientationVertical = false;
+
+        /// <summary>
+        /// If true, the GameObjectToPlace will remain upright and in line with Vector3.up
+        /// </summary>
+        public bool KeepOrientationVertical
+        {
+            get => keepOrientationVertical;
+            set
+            {
+                if (keepOrientationVertical != value)
+                {
+                    keepOrientationVertical = value;
+                }
+            }
+        }
+
+
+        [SerializeField]
+        [Tooltip("If true, the spatial mesh will be visible while the object is in the placing state.")]
+        private bool spatialMeshVisible = true;
+
+        /// <summary>
+        /// If true, the spatial mesh will be visible while the object is in the placing state.
+        /// </summary>
+        public bool SpatialMeshVisible
```

Comment:
Ah... this one is interesting.

There are a few things to mention here:

1. If the spatial awareness system is disabled or isn't configured to run, we should make sure that this code is resilient to this and doesn't crash (i.e. access to spatialAwarenessSystem doesn't get rekt)
2. We should be clear via documentation that this does have a dependency on the spatial system, and that it also controls the visibility of the mesh by default.

Also, what can be interesting is that there's no "override" for spatial system visibility. That is, imagine that you're an app that's consuming TapToPlace and spatialawareness at the same time - imagine that your app wants to actually have spatial awareness on at all times.

Now you pull in tap to place, and after tap to place has finished its work, it will disable/hide the spatial map! That's not great, because you wanted it to always show, but this other component (this thing) just disabled it for you.

This is a general problem in other areas of eventing/true/false enable control where you need to multiplex access (i.e. you want to have multiple people say they "want it enable or disabled" and if anyone wants it enabled, it will be enabled, and if no one wants it enabled, it will be disabled).

That's more complicated, but I think that in this case it might just be good to, when tap to place is done, "restore" the previous state of the mesh visibility (i.e. if it was disabled, make sure it gets disabled, it it was enabled, make sure it doens't get disabled).

This wouldn't solve the problem of:

1. Tap to place starts
2. Other app code enables mesh visibilty
3. Tap to place ends.

But it would at least fix the main issue.

It might be intersting to call out this issue in the comments (and suggest that consumers can set this to false and control the spatial mesh visibility as part of using tap to place?)

Reply:
Oh wait you're actually doing this with InitialSpatialMeshDisplayOption.

In that case, I think still calling out the unsolvable problem (i.e. the list mentioned above) would still be helpful.

Reply:
Why would we want TapToPlace depend on the spatial awareness system? I would expect to use this component even if I'm not using SR, just that it would place things on the bubble, or collide with other stuff in the scene...

Reply:
hello. I want (and expect) this script not to depend on SRat all. When spatial awareness is disabled in MR toolkit, gameobject with taptoplace script on it start to behave weirdly, go back and forth across the scene, start to rotate, jump from one position to another, etc. Still researching the reason. Actually i'm using spatial awareness, but at the moment of placing object i already have mesh of walls and floors and spatial mesh disabled. I hope, that i'll find out the reason of this behaviour, despite this, i found taptoplace script pretty good and useful

Reply:
@julenka The reason tap to place uses the spatial awareness system is to allow for the option to show the wireframe representation of the spatial mesh while the object is in the placing state.  Showing the spatial mesh while the object is in the placing state is behavior from HoloToolkit's tap to place.

I can add a check that looks to see if the spatial awareness system is enabled or I can remove the mesh display behavior completely. 

DA-NDI, thank you for testing!  The weird behavior you are seeing will be addressed.

Reply:
@CDiaz-MS "Showing the spatial mesh while the object is in the placing state is behavior from HoloToolkit's tap to place." 

I would also vote on removing this in the MRTK version of TapToPlace. It will simplify the code and remove dependencies. 

It is important to add events though in TapToPlace on when state transitions occur. This way people can enable or disable the SA system themselves if they so wish. 

We could also look at creating an extended TapToPlace component in the future that does this SA toggling functionality

Reply:
I really like the idea of removing the SA dependency and adding an IsBeingPlaced event.  
@julenka @wiwei What do you think about changing the original HTK behavior for tap to place MRTK?

Reply:
So the SA dependency is pretty neat as an option - I think an interesting immediate scenario/selling point is being able to turn it on and being able to place things on walls, rather than having to precreate geometry in the scene.

However, I think it might be more common for folks to load up the example in editor (and thus, having the geometry there and not have it be SA bound would probably be more immediately useful).

I think it would actually be a nice option to have (i.e. engage SA system during placement) but it's probably fine without. It also sidesteps the question of how to coordinate SA disable/enable (i.e. the multiplexing issue that I mentioned before)